core/src/chunk.rs

```diff
@@ -151,6 +213,58 @@ impl Chunk {
         }
 
         section.set_block_at(x, y % 16, z, block);
+
+        self.update_heightmap(x, y, z, block);
+    }
+
+    pub fn heightmap(&self, x: usize, z: usize) -> &HeightMap {
+        Self::check_coords(x, 0, z);
+        &self.heightmaps[x + z * CHUNK_WIDTH]
+    }
+
+    pub fn heightmap_mut(&mut self, x: usize, z: usize) -> &mut HeightMap {
+        Self::check_coords(x, 0, z);
+        &mut self.heightmaps[x + z * CHUNK_WIDTH]
+    }
+
+    pub fn heightmaps(&self) -> &[HeightMap] {
+        &self.heightmaps
+    }
+
+    fn update_heightmap(&mut self, x: usize, y: usize, z: usize, block: Block) {
+        let heightmap = self.heightmap_mut(x, z);
+        if (block.is_solid() || block.is_fluid()) && heightmap.motion_blocking() < y {
+            heightmap.set_motion_blocking(y);
+        }
+
+        if (block.is_solid() || block.is_fluid())
+            && !block.is_leaves()
+            && heightmap.motion_blocking_no_leaves() < y
+        {
+            heightmap.set_motion_blocking_no_leaves(y);
+        }
+
+        if block.is_solid() && heightmap.ocean_floor() < y {
+            heightmap.set_ocean_floor(y);
+        }
+
+        if !block.is_air() && heightmap.world_surface() < y {
+            heightmap.set_world_surface(y);
+        }
+    }
+
+    /// Recalculate the heightmap for the chunk
+    pub fn recalculate_heightmap(&mut self) {
+        // This function can be optimized, instead of 
+        // fetching heightmap every time, and sections
+        for x in 0..CHUNK_WIDTH {
+            for z in 0..CHUNK_WIDTH {
+                for y in (0..CHUNK_HEIGHT).rev() {
+                    let block = self.block_at(x, y, z);
+                    self.update_heightmap(x, y, z, block);
```

Comment:
Could we have `update_heightmap` return a bool that indicates whether it encountered a block? This way, we could skip checking the rest of the y values because we found the highest block

Reply:
Sure, that would be nice; however, still need to figure out how to recalculate the world generated heights, ie. `world_surface_wg` and `ocean_floor_wg`.